### ISSUE TITLE
Improve memoize

### ISSUE THREAD

**Issue by Technologicat** [Technologicat/unpythonic#78]
The differences of `unpythonic.memoize` and the nearest stdlib function [`functools.lru_cache`](https://github.com/python/cpython/blob/main/Lib/functools.py) need to be documented. They are:

- `memoize` binds arguments like Python itself does, so given this definition:

  ```python
  def f(a, b):
      return a, b
  ```

  and memoizing it, the calls `f(1, 2)`, `f(1, b=2)`, `f(a=1, b=2)`, and `f(b=2, a=1)` all hit the same cache key.

  As of Python 3.9, in `functools.lru_cache` this is not so; see the internal function `functools._make_key`, where the comments explicitly say so.

- `memoize` caches exceptions, too. A pure function that crashed for some combination of arguments will, if given the same inputs, crash again with the same error, so there is no reason to run it again.

- `memoize` does **not** have a `typed` mode to treat `42` and `42.0` as different keys. The keys are hashed, and both an `int` and an equal `float` hash to the same value.

  What the `typed` mode of `functools.lru_cache` is doing is actually a form of dispatch. Hence, you can use `@generic`, and `@memoize` each individual multimethod:

  ```python
  from unpythonic import generic, memoize

  @generic
  @memoize
  def thrice(x: int):
      return 3 * x

  @generic
  @memoize
  def thrice(x: float):
      return 3.0 * x
  ```

  If you do not want to opt in to `unpythonic`'s multiple-dispatch subsystem, you can manually separate your dispatcher and implementations into separate functions, and then **memoize each implementation** (not the dispatcher):

  ```python
  from unpythonic import memoize

  def thrice(x):  # the dispatcher
      if isinstance(x, int):
          return thrice_int(x)
      elif isinstance(x, float):
          return thrice_float(x)
      raise TypeError(type(x))

  @memoize
  def thrice_int(x):
      return 3 * x

  @memoize
  def thrice_float(x):
      return 3.0 * x
  ```

  These solutions keep dispatching and memoization orthogonal.

- `memoize` has no maximum cache size or hit/miss statistics counting.

- `memoize` is **not** thread-safe. This is a **BUG** that needs to be fixed ASAP.

**Comment by Technologicat** [comment]
Docs updated as of cce01e3.

Now all that remains for this issue is to fix the thread-safety bug.

**Comment by Technologicat** [comment]
Check also whether `fix` and `fixtco` have the same thread-safety bug, or if I fixed it there already earlier.

**Comment by Technologicat** [comment]
Bug fixed in f1551f4.

The `fix` and `fixtco` functions didn't have this bug; there each call chain has its own memo.
